File: algorithm/力扣热题100/17_电话号码的字母组合.js
/**
 * @param {string} digits
 * @return {string[]}
 */
var letterCombinations = function (digits) {
    const map = {
        2: 'abc'.split(''),
        3: 'def'.split(''),
        4: 'ghi'.split(''),
        5: 'jkl'.split(''),
        6: 'mno'.split(''),
        7: 'pqrs'.split(''),
        8: 'tuv'.split(''),
        9: 'wxyz'.split(''),
    }

    const res = []
    const path = []
    function bt(start) {
        if (start >= digits.length) {
            if (path.length > 0) {
                res.push(path.join(''))
            }
            return
        }

        const letters = map[digits[start]];
        for (let j = 0; j < letters.length; j++) {
            path.push(letters[j])
            bt(start + 1)
            path.pop()
        }
    }
    bt(0)
    return res;
};

File: algorithm/力扣热题100/17_电话号码的字母组合.go
package main

func letterCombinations(digits string) []string {
	letterMap := map[byte]string{
		'2': "abc",
		'3': "def",
		'4': "ghi",
		'5': "jkl",
		'6': "mno",
		'7': "pqrs",
		'8': "tuv",
		'9': "wxyz",
	}

	path := []byte{}
	res := []string{}
	var bt func(start int)
	bt = func(start int) {
		if start >= len(digits) {
			res = append(res, string(path))
			return
		}

		cur := letterMap[digits[start]]
		for i := 0; i < len(cur); i++ {
			path = append(path, cur[i])
			bt(start + 1)
			path = path[:len(path)-1]
		}
	}
	bt(0)
	return res
}

// func main() {
// 	// findMedianSortedArrays([]int{1, 3}, []int{2})
// 	fmt.Println(letterCombinations("23"))
// }
